File: codility/stacks_queues/brackets.py
def solution(S):
    stack = []
    open_close = {'{' : '}', '[':']', '(':')'}
    for car in S:
        if car in open_close.keys():
            stack.append(car)
        if car in open_close.values():
            if not stack or open_close[stack.pop()] != car:
                return 0
    if len(stack) == 0:
        return 1
    else:
        return 0


print(solution("{[()()]}"))
print(solution("([)()]"))

File: java/src/Brackets.java
import java.util.Map;
import java.util.Stack;

public class Brackets{
    public static boolean solution(String s){
        Stack<Character> chars = new Stack<>();
        Map<Character, Character> brackets = Map.of('{','}','(',')', '[',']');
        for(Character c : s.toCharArray()){
            if(brackets.keySet().contains(c)){
                chars.push(c);
            }
            else if(!chars.empty() && brackets.containsValue(c)){
                Character popped = chars.pop();
                if(c != brackets.get(popped)){
                    return false;
                }
            }
        }
        return chars.size() == 0;
    }
    public static void main(String[] args) {
        System.out.println(solution("{[()()]}"));
        System.out.println(solution("([)()]"));

    }
}
